src/Microsoft.Identity.Web.Certificate/SignedAssertionFromManagedIdentityCredentialLoader.cs

```diff
@@ -23,7 +19,7 @@ public async Task LoadIfNeededAsync(CredentialDescription credentialDescription,
                 ManagedIdentityClientAssertion? managedIdentityClientAssertion = credentialDescription.CachedValue as ManagedIdentityClientAssertion;
```

Comment:
It's not readily evident to me how .CachedValue is set, and how the new CredentialDescription.TokenExchangeUrl property is accounted for - and how does it map to ManagedIdentityClientAssertion? Is there a good starting point in the source code to see where this happens? @jmprieur @jennyf19

Reply:
@Kelly, here is how it works:

When developers add token acquisition (services.AddTokenAcquisition), it [adds](https://github.com/AzureAD/microsoft-identity-web/blob/256bf5edafc555756621151435661e11946986ce/src/Microsoft.Identity.Web.TokenAcquisition/ServiceCollectionExtensions.cs#L48) the [DefaultCertificateLoader](https://github.com/AzureAD/microsoft-identity-web/blob/256bf5edafc555756621151435661e11946986ce/src/Microsoft.Identity.Web.Certificate/DefaultCertificateLoader.cs#L28) service which inherits from [DefaultCredentialsLoader](https://github.com/AzureAD/microsoft-identity-web/blob/256bf5edafc555756621151435661e11946986ce/src/Microsoft.Identity.Web.Certificate/DefaultCredentialsLoader.cs#L16).

The default certificate/credential loader is called by IdWeb passing a [CredentialDescription](https://github.com/AzureAD/microsoft-identity-abstractions-for-dotnet/blob/main/src/Microsoft.Identity.Abstractions/ApplicationOptions/CredentialDescription.cs) (or a collection of CredentialDescriptions), and it chooses the [right ](https://github.com/AzureAD/microsoft-identity-web/blob/256bf5edafc555756621151435661e11946986ce/src/Microsoft.Identity.Web.Certificate/DefaultCredentialsLoader.cs#L35 )ICredentialSourceLoader based on the [CredentialSource](url) property in the CredentialDescription.

When it's a SignedAssertionFromManagedIdentity it [creates](https://github.com/AzureAD/microsoft-identity-web/blob/256bf5edafc555756621151435661e11946986ce/src/Microsoft.Identity.Web.Certificate/SignedAssertionFromManagedIdentityCredentialLoader.cs#L26) a ManagedIdentityClientAssertion based on the information in the CredentialDescription.

In Microsoft.Identity.Web.Abstractions, you've done the work to add a property to the CredentialDescription, and it's magically available (after reading the appsetting.json) in SignedAssertionFromManagedIdentity where you processed it.

See also https://github.com/AzureAD/microsoft-identity-web/wiki/v2.0#common-configuration

Reply:
^Really appreciate this, JM. Thank you!!!!